tensorflow_probability/python/experimental/mcmc/with_reductions_test.py

```diff
@@ -0,0 +1,491 @@
+    self.assertEqual(pkr.streaming_calculations, None)
+    self.assertEqual(inner_pkr.counter_1, 0)
+    self.assertEqual(inner_pkr.counter_2, 0)
+
+  def test_is_calibrated(self):
+    fake_calibrated_kernel = TestTransitionKernel()
+    fake_uncalibrated_kernel = TestTransitionKernel(is_calibrated=False)
+    fake_reducer = TestReducer()
+    calibrated_reducer_kernel = WithReductions(
+        inner_kernel=fake_calibrated_kernel,
+        reducers=fake_reducer,
+    )
+    uncalibrated_reducer_kernel = WithReductions(
+        inner_kernel=fake_uncalibrated_kernel,
+        reducers=fake_reducer,
+    )
+    self.assertTrue(calibrated_reducer_kernel.is_calibrated)
+    self.assertFalse(uncalibrated_reducer_kernel.is_calibrated)
+
+  def test_tf_while(self):
+    fake_kernel = TestTransitionKernel()
+    fake_reducer = TestReducer()
+    reducer_kernel = WithReductions(
+        inner_kernel=fake_kernel,
+        reducers=fake_reducer,
+    )
+    initial_sample, initial_kernel_results = reducer_kernel.one_step(0.,)
+
+    def _loop_body(i, curr_state, pkr):
+      new_state, kernel_results = reducer_kernel.one_step(curr_state, pkr)
+      return (i + 1, new_state, kernel_results)
+    _, new_sample, kernel_results = tf.while_loop(
+        lambda i, _, __: i < 5,
+        _loop_body,
+        (0., initial_sample, initial_kernel_results)
+    )
+
+    new_sample, kernel_results = self.evaluate([
+        new_sample, kernel_results])
+    self.assertEqual(kernel_results.streaming_calculations, 6)
+    self.assertEqual(new_sample, 6)
+    self.assertEqual(kernel_results.inner_results.counter_1, 6)
+    self.assertEqual(kernel_results.inner_results.counter_2, 12)
+
+  def test_nested_reducers(self):
+    fake_kernel = TestTransitionKernel()
+    nested_reducers = [[TestReducer(), TestReducer()], [TestReducer()]]
+    reducer_kernel = WithReductions(
+        inner_kernel=fake_kernel,
+        reducers=nested_reducers,
+    )
+    new_sample, kernel_results = reducer_kernel.one_step(0.,)
+    new_sample, kernel_results = self.evaluate([
+        new_sample, kernel_results])
+
+    self.assertEqual(
+        len(kernel_results.streaming_calculations[0]), 2)
+    self.assertEqual(
+        len(kernel_results.streaming_calculations[1]), 1)
+    self.assertEqual(
+        np.array(kernel_results.streaming_calculations).shape,
+        (2,))
+
+    self.assertAllEqual(
+        kernel_results.streaming_calculations,
+        [[1, 1], [1]])
+    self.assertEqual(new_sample, 1)
+    self.assertEqual(kernel_results.inner_results.counter_1, 1)
+    self.assertEqual(kernel_results.inner_results.counter_2, 2)
+
+  def test_nested_state_dependent_reducers(self):
+    fake_kernel = TestTransitionKernel()
+    nested_reducers = [[MeanReducer(), MeanReducer()], [MeanReducer()]]
+    reducer_kernel = WithReductions(
+        inner_kernel=fake_kernel,
+        reducers=nested_reducers,
+    )
+    new_sample, kernel_results = reducer_kernel.one_step(0.,)
```

Comment:
My other review comments may required manually bootstrapping results in this test and others.